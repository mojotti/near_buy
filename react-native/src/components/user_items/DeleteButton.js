import React, { Component } from 'react';
import { Alert, Text, TouchableHighlight } from 'react-native';
import PropTypes from 'prop-types';

import { styles } from '../../static/styles/UserItemDetailsStyles';
import { DELETION_ERROR, localhost } from '../../static/constants';
import { getBearerHeaders } from '../../networking/networking';

export default class DeleteButton extends Component {
  _navigateBack = () => {
    const { goBack } = this.props.navigation;
    goBack();
  };

  _removeItemAndNavigateBack = () => {
    const url = `http://${localhost}:5000/api/v1.0/user/items/${this.props.id}`;
    fetch(url, {
      method: 'DELETE',
      headers: getBearerHeaders(this.props.token),
    })
      .then(response => response.json())
      .then(responseJson => {
        if (responseJson.ok) {
          this.props.fetchItems();
          this._navigateBack();
        }
      })
      .catch(error => {
        Alert.alert(...DELETION_ERROR);
        console.error(error);
      });
  };

  render() {
    return (
      <TouchableHighlight
        onPress={this._removeItemAndNavigateBack}
        style={styles.deleteButton}
      >
        <Text style={styles.buttonText}>Delete item</Text>
      </TouchableHighlight>
    );
  }
}

DeleteButton.propTypes = {
  id: PropTypes.number.isRequired,
  token: PropTypes.string.isRequired,
};
