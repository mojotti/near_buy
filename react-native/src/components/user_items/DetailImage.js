import React, { Component } from 'react';
import { Dimensions, View } from 'react-native';
import ImageLoad from 'react-native-image-placeholder';
import MaterialIcons from 'react-native-vector-icons/MaterialIcons';
const { height } = Dimensions.get('window');

export default class DetailImage extends Component {
  renderImage = () => {
    return (
      <View style={{ flex: 1 }}>
        <ImageLoad
          style={{ flex: 1 }}
          placeholderStyle={{
            flex: 1,
            alignSelf: 'center',
            alignItems: 'center',
            justifyContent: 'center',
          }}
          source={{ uri: this.props.url }}
          isShowActivity={false}
        />
      </View>
    );
  };

  renderPlaceholder = () => {
    return (
      <View
        style={{
          flex: 1,
          alignSelf: 'center',
          alignItems: 'center',
          justifyContent: 'center',
          paddingBottom: 40,
          backgroundColor: '#FFFFFF',
        }}
      >
        <MaterialIcons name="add-a-photo" size={height * 0.4} color="gray" />
      </View>
    );
  };

  render() {
    return this.props.url ? this.renderImage() : this.renderPlaceholder();
  }
}
