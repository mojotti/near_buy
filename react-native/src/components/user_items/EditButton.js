import React, { Component } from 'react';
import { Alert, Text, TouchableHighlight } from 'react-native';
import PropTypes from 'prop-types';

import { styles } from '../../static/styles/UserItemDetailsStyles';
import { DELETION_ERROR, localhost } from '../../static/constants';
import { getHeadersForRegistering } from '../../networking/networking';

export default class EditButton extends Component {
  _getHeaders = () => {
    const headers = new Headers();
    headers.append('Authorization', `Bearer ${this.props.token}`);
    headers.append('Content-Type', 'application/json');
    return headers;
  };

  _navigateBack = () => {
    const { goBack } = this.props.navigation;
    goBack();
  };

  _updateItemAndNavigateBack = () => {
    const url = `http://${localhost}:5000/api/v1.0/items/${this.props.id}`;
    console.log(this.props.location);
    fetch(url, {
      method: 'PUT',
      headers: this._getHeaders(),
      body: JSON.stringify({
        id: this.props.item.id,
        title: this.props.item.title,
        description: this.props.item.description,
        price: this.props.item.price,
        longitude: this.props.location.longitude,
        latitude: this.props.location.latitude,
        sold: false,
      }),
      N,
    })
      .then(response => response.json())
      .then(responseJson => {
        console.log('response', responseJson);
        this.props.fetchItems();
        this._navigateBack();
      })
      .catch(error => {
        Alert.alert(...DELETION_ERROR);
        console.error(error);
      });
  };

  render() {
    return (
      <TouchableHighlight
        onPress={this._updateItemAndNavigateBack}
        style={styles.editButton}
      >
        <Text style={styles.buttonText}>Edit item</Text>
      </TouchableHighlight>
    );
  }
}

EditButton.propTypes = {
  id: PropTypes.number.isRequired,
  token: PropTypes.string.isRequired,
};
