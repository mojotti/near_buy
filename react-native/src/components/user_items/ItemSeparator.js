import React, { Component } from 'react';
import { View } from 'react-native';
import { styles } from '../../static/styles/ListViewItemStyle';

export default class ItemSeparator extends Component {
  render() {
    return <View style={styles.separator} />;
  }
}
