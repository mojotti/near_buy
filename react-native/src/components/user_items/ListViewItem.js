import React from 'react';
import { Image, Text, View } from 'react-native';
import { styles } from '../../static/styles/ListViewItemStyle';
import { localhost } from '../../static/constants';

export class ListViewItem extends React.Component {
  render() {
    const imagePath = `http://${localhost}:5000/${
      this.props.itemId
    }/image0.jpg`;

    return (
      <View style={{ flexDirection: 'row', flex: 1 }}>
        <Image
          style={{ width: 50, height: 50 }}
          source={{
            uri: imagePath,
          }}
        />
        <View style={[styles.item, styles.separator]}>
          <Text>{`Title: ${this.props.item.title}`}</Text>
          <Text>{`Description: ${this.props.item.description}`}</Text>
          <Text>{`Price: ${this.props.item.price} â‚¬`}</Text>
        </View>
      </View>
    );
  }
}
