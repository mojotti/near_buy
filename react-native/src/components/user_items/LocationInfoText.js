import React from 'react';
import { Text, View } from 'react-native';
import { styles } from '../../static/styles/UserItemDetailsStyles';

export default class LocationInfoText extends React.Component {
  render() {
    return (
      <View style={{ alignSelf: 'stretch' }}>
        <Text
          style={styles.infoText}
        >{`\u2022 Your exact location is not visible to others`}</Text>
        <Text style={styles.infoText}>
          {`\u2022`}
          <Text style={{ color: 'blue' }}> Blue</Text>
          <Text>{` color indicates your item's location`}</Text>
        </Text>
        <Text style={styles.infoText}>
          {`\u2022`}
          <Text style={{ color: 'red' }}> Red</Text>
          <Text>{` color indicates your location`}</Text>
        </Text>
      </View>
    );
  }
}
