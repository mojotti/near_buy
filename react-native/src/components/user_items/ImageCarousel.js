import React from 'react';
import PropTypes from 'prop-types';
import Carousel from 'react-native-looped-carousel';
import { styles } from '../../static/styles/UserItemDetailsStyles';
import DetailImage from './DetailImage';

export default class ImageCarousel extends React.Component {
  render() {
    return (
      <Carousel
        style={styles.carousel}
        autoplay={false}
        pageInfo
        currentPage={0}
        pageInfoBackgroundColor={'#292929'}
        pageInfoTextStyle={{ color: '#FFFFFF' }}
    // arrows are disabled for now since they don't work as intended on ios
        arrows={false}
        arrowStyle={styles.arrowStyle}
        leftArrowText={' '}
        rightArrowText={' '}
      >
        <DetailImage
          url={this.props.imageUrls[0]}
          id={this.props.id}
          numOfPics={this.props.numOfPics}
          onImageUpload={this.props.onImageUpload}
        />
        <DetailImage
          url={this.props.imageUrls[1]}
          id={this.props.id}
          numOfPics={this.props.numOfPics}
          onImageUpload={this.props.onImageUpload}
        />
        <DetailImage
          url={this.props.imageUrls[2]}
          id={this.props.id}
          numOfPics={this.props.numOfPics}
          onImageUpload={this.props.onImageUpload}
        />
        <DetailImage
          url={this.props.imageUrls[3]}
          id={this.props.id}
          numOfPics={this.props.numOfPics}
          onImageUpload={this.props.onImageUpload}
        />
      </Carousel>
    );
  }
}

ImageCarousel.propTypes = {
  imageUrls: PropTypes.arrayOf(PropTypes.string).isRequired,
  id: PropTypes.number.isRequired,
  onImageUpload: PropTypes.func.isRequired,
  numOfPics: PropTypes.number.isRequired,
};
