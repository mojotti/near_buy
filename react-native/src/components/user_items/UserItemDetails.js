import React, { Component } from 'react';
import { ScrollView, Text, View } from 'react-native';
import { connect } from 'react-redux';

import { localhost } from '../../static/constants';
import { styles } from '../../static/styles/UserItemDetailsStyles';
import { ItemDetails } from '../new_item/ItemDetails';
import DeleteButton from './DeleteButton';
import SaveButton from './SaveButton';
import UserItemMapView from './UserItemMapView';
import { getBearerHeaders } from '../../networking/networking';
import ImageCarousel from './ImageCarousel';

class _UserItemDetails extends Component {
  constructor(props) {
    super(props);
    const { params } = this.props.navigation.state;
    const item = params ? params.item : null;

    this.state = {
      id: item.id,
      title: item.title,
      description: item.description,
      price: item.price,
      created: item.item_created,
      imageUrls: [],
      longitude: item.longitude,
      latitude: item.latitude,
    };

    this.fetchItems = params.fetchItems;
  }

  static navigationOptions = ({ navigation }) => {
    const { params } = navigation.state;
    const item = params ? params.item : null;

    return {
      title:
        typeof navigation.state.params === 'undefined' ||
        typeof navigation.state.params.title === 'undefined'
          ? ''
          : navigation.state.params.title,
      headerRight: (
        <SaveButton
          id={item.id}
          navigation={navigation}
          fetchItems={params.fetchItems}
          item={item}
        />
      ),
    };
  };

  componentDidMount() {
    this.getNumOfPictures().then(numOfPics => this.getImageUrls(numOfPics));
    this.props.navigation.setParams({ title: this.state.title });
  }

  componentDidUpdate(prevProps, prevState) {
    if (
      prevState.title !== this.state.title ||
      prevState.description !== this.state.description ||
      prevState.price !== this.state.price ||
      prevState.longitude !== this.state.longitude ||
      prevState.latitude !== this.state.latitude
    ) {
      this.props.navigation.setParams({ item: this.state });
    }
  }

  getNumOfPictures = () => {
    const url = `http://${localhost}:5000/api/v1.0/${
      this.props.navigation.state.params.item.id
    }/num_of_images`;
    return fetch(url, {
      method: 'GET',
      headers: getBearerHeaders(this.props.token),
    })
      .then(response => response.json())
      .then(responseJson => {
        return responseJson.num_of_images;
      })
      .catch(error => {
        console.error(error);
        return 0;
      });
  };

  getImageUrls = numOfImages => {
    for (let i = 0; i < numOfImages; i++) {
      const imagePath = `http://${localhost}:5000/api/v1.0/${
        this.state.id
      }/image${i}.jpg`;

      this.setState(prevState => {
        let images = prevState.imageUrls;
        images[i] = imagePath;
        return images;
      });
    }
  };

  handleTitleChange = text => {
    this.setState({ title: text });
  };

  handleDescriptionChange = text => {
    this.setState({ description: text });
  };

  handlePriceChange = text => {
    this.setState({ price: text });
  };

  render() {
    const itemDetailsProps = {
      onTitleChange: this.handleTitleChange,
      onDescriptionChange: this.handleDescriptionChange,
      onPriceChange: this.handlePriceChange,
      title: this.state.title,
      description: this.state.description,
      price: this.state.price,
    };

    return (
      <ScrollView keyboardShouldPersistTaps="handled">
        <View style={styles.container}>
          <Text style={styles.headerText}>Pictures</Text>
          <ImageCarousel imageUrls={this.state.imageUrls} />
          <Text style={styles.headerText}>Edit details</Text>
          <ItemDetails {...itemDetailsProps} />
          <Text style={styles.headerText}>Location</Text>
          <Text
            style={styles.infoText}
          >{`Your exact location is not visible to others`}</Text>
          <UserItemMapView
            longitude={this.state.longitude}
            latitude={this.state.latitude}
            currentLocation={this.props.location}
          />
          <Text style={styles.headerText}>Fun facts</Text>
          <Text
            style={styles.paragraph}
          >{`\u2022 You decided to get rid of this item ${new Date(
            this.state.created,
          ).toDateString()}`}</Text>
          <DeleteButton
            id={this.state.id}
            token={this.props.token}
            navigation={this.props.navigation}
            fetchItems={this.fetchItems}
          />
        </View>
      </ScrollView>
    );
  }
}

const mapStateToProps = state => {
  const { latitude, longitude } = state.locationReducer;
  return {
    token: state.authorizationReducer.token,
    location: {
      latitude,
      longitude,
    },
  };
};

const UserItemDetails = connect(mapStateToProps, null)(_UserItemDetails);
export default UserItemDetails;
