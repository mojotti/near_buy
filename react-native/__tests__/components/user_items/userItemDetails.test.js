import React from 'react';
import { shallow } from 'enzyme';
import { _UserItemDetails } from '../../../src/components/user_items/UserItemDetails';


let setParamsSpy = null;
let PROPS = null;

const getProps = paramsSpy => {
  return {
    navigation: {
      state: {
        params: {
          item: {
            id: 0,
            title: 'foo',
            description: 'bar',
            price: 100,
            item_created: 103042050,
            longitude: 25,
            latitude: 65.02,
          },
          fetchItems: () => {},
          isEdited: true,
        },
      },
      setParams: paramsSpy,
    },
    location: {
      longitude: 24,
      latitude: 65,
    },
    token: 'foo',
  };
};

describe('<_UserItemDetails />', () => {
  beforeEach(() => {
    setParamsSpy = jest.fn();
    PROPS = getProps(setParamsSpy);
  });

  test('renders correctly', () => {
    fetch.mockResponseSuccess({ num_of_images: 1 });
    const userItemDetails = shallow(<_UserItemDetails {...PROPS} />);

    expect(userItemDetails).toMatchSnapshot();
  });

  test('component did update', () => {
    fetch.mockResponseSuccess({ num_of_images: 1 });
    const userItemDetails = shallow(<_UserItemDetails {...PROPS} />);
    expect(setParamsSpy.mock.calls.length).toEqual(1);

    const prevState = {
      title: 'bar',
      price: 450,
    };
    userItemDetails.instance().componentDidUpdate({}, prevState);
    expect(setParamsSpy.mock.calls.length).toEqual(2);
  });

  test('state handlers', () => {
    fetch.mockResponseSuccess({ num_of_images: 1 });

    const userItemDetails = shallow(<_UserItemDetails {...PROPS} />);

    expect(userItemDetails.state('title')).toEqual('foo');
    expect(userItemDetails.state('description')).toEqual('bar');
    expect(userItemDetails.state('price')).toEqual(100);
    expect(userItemDetails.state('latitude')).toEqual(65.02);
    expect(userItemDetails.state('longitude')).toEqual(25);

    userItemDetails.instance().handleTitleChange('title');
    userItemDetails.instance().handleDescriptionChange('description');
    userItemDetails.instance().handlePriceChange(250);
    userItemDetails.instance().updateLocation();

    expect(userItemDetails.state('title')).toEqual('title');
    expect(userItemDetails.state('description')).toEqual('description');
    expect(userItemDetails.state('price')).toEqual(250);
    expect(userItemDetails.state('latitude')).toEqual(65);
    expect(userItemDetails.state('longitude')).toEqual(24);
  });

  test('navigation options', () => {
    const navigation = {
      state: {
        params: {
          item: {
            id: 0,
            title: 'foo',
            description: 'bar',
            price: 100,
            item_created: 103042050,
            longitude: 25,
            latitude: 65.02,
          },
          fetchItems: () => {},
          isEdited: true,
        },
      },
    };

    const params = { navigation };
    expect(_UserItemDetails.navigationOptions(params)).toMatchSnapshot();
  });
});
