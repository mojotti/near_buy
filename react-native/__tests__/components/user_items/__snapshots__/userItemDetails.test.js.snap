// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<_UserItemDetails /> navigation options 1`] = `
Object {
  "headerRight": <Connect(_SaveButton)
    fetchItems={[Function]}
    id={0}
    item={
      Object {
        "description": "bar",
        "id": 0,
        "item_created": 103042050,
        "latitude": 65.02,
        "longitude": 25,
        "price": 100,
        "title": "foo",
      }
    }
    navigation={
      Object {
        "state": Object {
          "params": Object {
            "fetchItems": [Function],
            "isEdited": true,
            "item": Object {
              "description": "bar",
              "id": 0,
              "item_created": 103042050,
              "latitude": 65.02,
              "longitude": 25,
              "price": 100,
              "title": "foo",
            },
          },
        },
      }
    }
  />,
  "title": "",
}
`;

exports[`<_UserItemDetails /> renders correctly 1`] = `
<ScrollView
  keyboardShouldPersistTaps="handled"
>
  <Component
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Component
      style={
        Object {
          "alignSelf": "stretch",
          "color": "#848484",
          "fontFamily": null,
          "fontSize": 25,
          "marginBottom": 15,
          "marginLeft": 56.25,
          "marginTop": 20,
          "textAlign": "left",
        }
      }
    >
      Pictures
    </Component>
    <ImageCarousel
      id={0}
      imageUrls={Array []}
      numOfPics={0}
      onImageUpload={[Function]}
    />
    <Component
      style={
        Object {
          "alignSelf": "stretch",
          "color": "#848484",
          "fontFamily": null,
          "fontSize": 25,
          "marginBottom": 15,
          "marginLeft": 56.25,
          "marginTop": 20,
          "textAlign": "left",
        }
      }
    >
      Edit details
    </Component>
    <ItemDetails
      description="bar"
      onDescriptionChange={[Function]}
      onPriceChange={[Function]}
      onTitleChange={[Function]}
      price={100}
      title="foo"
    />
    <Component
      style={
        Object {
          "alignSelf": "stretch",
          "color": "#848484",
          "fontFamily": null,
          "fontSize": 25,
          "marginBottom": 15,
          "marginLeft": 56.25,
          "marginTop": 20,
          "textAlign": "left",
        }
      }
    >
      Location
    </Component>
    <LocationInfoText />
    <UserItemMapView
      currentLocation={
        Object {
          "latitude": 65,
          "longitude": 24,
        }
      }
      latitude={65.02}
      longitude={25}
    />
    <UpdateLocationButton
      updateLocation={[Function]}
    />
    <Component
      style={
        Object {
          "alignSelf": "stretch",
          "color": "#848484",
          "fontFamily": null,
          "fontSize": 25,
          "marginBottom": 15,
          "marginLeft": 56.25,
          "marginTop": 20,
          "textAlign": "left",
        }
      }
    >
      Fun facts
    </Component>
    <Component
      style={
        Object {
          "alignSelf": "center",
          "fontFamily": null,
          "marginBottom": 10,
          "textAlign": "center",
        }
      }
    >
      â€¢ You decided to get rid of this item Fri Jan 01 2016
    </Component>
    <DeleteButton
      fetchItems={[Function]}
      id={0}
      navigation={
        Object {
          "setParams": [Function],
          "state": Object {
            "params": Object {
              "fetchItems": [Function],
              "isEdited": true,
              "item": Object {
                "description": "bar",
                "id": 0,
                "item_created": 103042050,
                "latitude": 65.02,
                "longitude": 25,
                "price": 100,
                "title": "foo",
              },
            },
          },
        }
      }
      token="foo"
    />
  </Component>
</ScrollView>
`;
